ivadomed/ivadomed
# Prototype Incorporation of Loader v2 with Alternative File based Loader Config JSON Support

## Checklist

#### GitHub




#### PR contents


## Description


For context, please see #1087 #1044 #1009 #1102

TLDR: 

* This introduce a optional LoaderV2 data loading path which COEXISTS with existing BIDS Based file indexing system, entirely triggered by three additional keywords in the config json files: 
  * `dataset_groups` which is a list of dictionaries. See example at `ivadomed/config/example_loader_v2_configs.py`
  * `expected_input` determines the channels of input to the model. This will overwrite any subdataset and filter out those without matching number of input. 
  * `expected_gt`  determines the channels of input to the model. This will overwrite any subdataset and filter out those without matching number of gt. 

* New Data Loader Classes at 
   * `ivadomed/loader/consolidation.py`   Consolidation class used to combine data from various dataset_groups etc. 
   * `ivadomed/loader/all_dataset_group.py`   Highest level data organization class used to represent ALL data used for training
   * `ivadomed/loader/dataset_group.py`  Mid level data organization class used to represent a set of train/val/test data used
   * `ivadomed/loader/files_dataset.py`  Low level data organization class used to represent a set of files. 
   * `ivadomed/loader/generalized_loader_configuration.py`  Aggregated configuration information to streamline IO. 

* `Train.py` and `Main.py` reorganization to support this optional analyses path. 
   
* Mocker: to help generate BIDS like mocking NII file structure utilized for testing. 
  * Located at `testing/mocker` module. 

* Schematics available at https://drive.google.com/file/d/1G69tgZzktcdbvSu48dCRL6QPVavZryxF/view?usp=sharing 
* Example JSON POC at https://drive.google.com/file/d/1-XrLOGc-4vW-H-px7rrDH1r2ca4aEdlZ/view?usp=sharing 

## Original Objectives of the POC: Aggregator and Centralized Config
  
    [X] Centralize Parameters to MRI2DSeg/3DSeg calls (~10 places)
    [X] Build JSON parser for File Path Type Dataset
    [X] Introduce Min/Max File Number validation mechanism
    [X] Introduce Incorrect File Number handling mechanism
    [X] Introduce Backward compatible Loader-Parameters key changes to existing JSON
    [ ] Update BIDS DataSet to incorporate the additional files
    [X] Build DataSetAggregator to concatenate the filename_pairs from File Path dataset types
  
## Other clean ups: 
* load_fileset.py can probably be removed. 

## Limitations: 

There are still MANY limitations that we aim to address in the future update:
* No metadata handling 
* Conflicting/unexpected files/dataset are dropped during aggregation process
* No train/val/test splitting automatically. Must manually specify the split. 

## How to Test:
Since this is a fully featured PR, it comes with its own unit test at `testing/functional_tests/test_training.py`. Beside that, it is fully transparently integrated with existing Ivadomed training process. The only update required are all on the JSON aspect so as long as the new keywords are properly prepared using Loader V2 format, everything else stays mostly the same for a basic training process in theory. See  `testing/functional_tests/test_training.py` (only ~30 lines) to get a sense of how the JSON is updated to enable the V2 loader training.  We expect test training to reveal more issues. 




## Linked issues


Partially address some of the problem of #1087 #1044 #1009 #1102
